S2A credentials and security connector API's.

matthewstevenson88 opened the pull request:
This PR contains the API's for the S2A credentials and the S2A security connector. Both sets of API's are nearly identical to their ALTS counterparts, namely [alts_credentials.h](https://github.com/grpc/grpc/blob/master/src/core/lib/security/credentials/alts/alts_credentials.h) and [alts_security_connector.h](https://github.com/grpc/grpc/blob/master/src/core/lib/security/security_connector/alts/alts_security_connector.h).

matthewstevenson88 commented:
The merge conflicts have been resolved and this PR is again ready for review. The modified files that require review are listed below:
- grpc_security.h
- grpc_s2a_credentials_options.{h,cc}
- s2a_credentials.{h,cc}
- s2a_security_connector.{h,cc}

matthewstevenson88 commented:
@yihuazhang I've rectified conflicts for this PR after several merges into s2a-master from other branches. PTAL. There is very little content - just API's for options, credentials, and security connector.

yihuazhang commented on include/grpc/grpc_security.h:
```diff
@@ -698,6 +698,64 @@ GRPCAPI grpc_channel_credentials* grpc_local_credentials_create(
+
+/** The main interface for the S2A credentials options. The options will contain
+ *  information that will be passed from gRPC to the TSI layer, such an ordered
+ *  list of supported ciphersuites or target service accounts. The S2A client
```
Will S2A also use target service accounts for strawman server authorization check? I think that feature is specific to ALTS.

yihuazhang commented on include/grpc/grpc_security.h:
```diff
@@ -698,6 +698,64 @@ GRPCAPI grpc_channel_credentials* grpc_local_credentials_create(
+ *  On success, it returns the created S2A channel credential object; otherwise,
+ *  it returns nullptr. **/
+GRPCAPI grpc_channel_credentials* grpc_s2a_credentials_create(
+    const grpc_s2a_credentials_options* options);
```
Please specify whether the ownership of `options` is transferred or not.

matthewstevenson88 replied:
The ownership should not be transferred. I have added a comment to this effect in the most recent commit.

yihuazhang commented on include/grpc/grpc_security.h:
```diff
@@ -698,6 +698,64 @@ GRPCAPI grpc_channel_credentials* grpc_local_credentials_create(
+ *
+ *  On success, it returns the created S2A server credential object; otherwise,
+ *  it returns nullptr. **/
+GRPCAPI grpc_server_credentials* grpc_s2a_server_credentials_create(
```
Same comment as above (i.e., on ownership).

matthewstevenson88 replied:
Same as above.

yihuazhang commented on include/grpc/grpc_security.h:
```diff
@@ -698,6 +698,64 @@ GRPCAPI grpc_channel_credentials* grpc_local_credentials_create(
+ *  information that will be passed from gRPC to the TSI layer, such an ordered
+ *  list of supported ciphersuites or target service accounts. The S2A client
+ *  (channel) and server credentials will have their own implementation of this
+ *  interface. The APIs listed in this header are thread-compatible. It is used
```
Just curious, how different are server and client's credentials options?

yihuazhang commented on src/core/lib/security/credentials/s2a/grpc_s2a_credentials_options.cc:
```diff
@@ -0,0 +1,64 @@
+void grpc_s2a_credentials_options_destroy(
+    grpc_s2a_credentials_options* options) {
+  if (options != nullptr) {
+    if (options->vtable != nullptr && options->vtable->destruct != nullptr) {
```
Now that gRPC C-core team is migrating towards C++, I am not sure if we want to keep using these C concept (e.g., v-table). F.Y.I. existing ALTS implementations need to be migrated to C++ as well.

matthewstevenson88 replied:
I've C++-ified the definition of this struct to make it look more like the `grpc_tls_credentials_options`struct that we use for SPIFFE.

yihuazhang commented on src/core/lib/security/credentials/s2a/grpc_s2a_credentials_options.h:
```diff
@@ -0,0 +1,93 @@
+/** An ordered list of ciphersuites for use by the S2A. **/
+typedef struct s2a_ciphersuite {
+  struct s2a_ciphersuite* next;
+  s2a_supported_ciphersuite cipher;
```
It seems a little overkill for using linked list here since we will only have three values at most. Can we use an array (of three elements) and keep track of its actual size?

matthewstevenson88 replied:
Good point. I replaced this with a `std::vector` as a member variable of the `grpc_s2a_credentials_options` class. (Note that we are allowed to use `std::vector` now in the C core.)

yihuazhang commented on src/core/lib/security/credentials/s2a/s2a_credentials.cc:
```diff
@@ -0,0 +1,65 @@
+
+grpc_s2a_credentials::grpc_s2a_credentials(
+    const grpc_s2a_credentials_options* options,
+    const char* handshaker_service_url)
```
Can we place `handshaker_service_url` inside `grpc_s2a_credentials_options`?

matthewstevenson88 replied:
Good call - added it as a member variable.

yihuazhang commented on src/core/lib/security/credentials/s2a/s2a_credentials.cc:
```diff
@@ -0,0 +1,65 @@
+
+grpc_s2a_server_credentials::grpc_s2a_server_credentials(
+    const grpc_s2a_credentials_options* options,
+    const char* handshaker_service_url)
```
Same as above.

matthewstevenson88 replied:
Done.

yihuazhang commented:
Sorry for the delay and looks good overall. My major concern is for `grpc_s2a_credentials_options` whether we should keep using C concept  (e.g., v-table) or not especially nowadays gRPC core is migrating towards C++.

matthewstevenson88 commented:
Thanks @yihuazhang! After looking at your comments, I've made a couple of changes:
1. I've C++-ified the `grpc_s2a_credentials_options` struct, making it look similar to the `grpc_tls_credentials_options` struct that we use for SPIFFE. Further, the linked lists are replaced with `std::vector`'s.
2. I've removed the distinction between "client options" and "server options".
3. I've added an "options test file" to illustrate how the `grpc_s2a_credentials_options` struct should be used.

Finally, it appears that there were a couple comments of yours that I received emails about but don't appear above. Did you delete them? Should I address them still?

yihuazhang commented on src/core/lib/security/credentials/s2a/grpc_s2a_credentials_options.cc:
```diff
@@ -22,43 +22,89 @@
+}
+
+void grpc_s2a_credentials_options::add_target_service_account(
+    const char* target_service_account) {
```
Please also do a sanity check on `target_service_account`

matthewstevenson88 replied:
Added a check to ensure `target_service_account` is not `nullptr`.

yihuazhang commented on src/core/lib/security/credentials/s2a/grpc_s2a_credentials_options.cc:
```diff
@@ -22,43 +22,89 @@
+  return new_options;
+}
+
+bool grpc_s2a_credentials_options::check_fields(
```
Is this API used anywhere other than tests? If not, can we make it a static method instead of class member function?

matthewstevenson88 replied:
Yes, it is used for testing purposes. I've added a comment to that effect.

yihuazhang commented on BUILD:
```diff
@@ -2035,21 +2037,21 @@ grpc_cc_library(
-        "grpc_base",
-    ],
-)
+#grpc_cc_library(
```
Maybe remove this rule, instead of commenting it out?

matthewstevenson88 replied:
Thanks! I forgot to delete it.

yihuazhang commented on src/core/lib/security/credentials/s2a/grpc_s2a_credentials_options.h:
```diff
@@ -21,73 +21,38 @@
+   * appropriate vector, if a duplicate exists. **/
+  void set_handshaker_service_url(const char* handshaker_service_url);
+  void add_supported_ciphersuite(uint16_t ciphersuite);
+  void add_target_service_account(const char* target_service_account);
```
Maybe adding comments saying that this API should be called only at the client side?

matthewstevenson88 replied:
Done.

yihuazhang commented:
LGTM and Approval

matthewstevenson88 commented:
Thanks for the review @yihuazhang! 

@cesarghali can you have a look now? The part I'd like your thoughts on is `grpc_s2a_credentials_options.{h,cc}`, and whether there are any tweaks to the options that need to be made. Note that these options will be used for 2 purposes:
1. Preparing the client/server start message.
2. Establishing the connection with the S2A service.

cesarghali commented on src/core/lib/security/credentials/s2a/s2a_credentials.h:
```diff
@@ -0,0 +1,68 @@
+constexpr char kGrpcS2ACredentialsType[] = "s2a";
+
+/** The main client-side struct for securing a gRPC connection using S2A. **/
+class grpc_s2a_credentials final : public grpc_channel_credentials {
```
Same comment RE adding "client" in the name.

matthewstevenson88 replied:
My same comment applies here: see [`alts_credentials.h`](https://github.com/grpc/grpc/blob/master/src/core/lib/security/credentials/alts/alts_credentials.h) or `spiffe_credentials.h`](https://github.com/grpc/grpc/blob/master/src/core/lib/security/credentials/tls/spiffe_credentials.h).

cesarghali commented on test/core/security/s2a_credentials_test.cc:
```diff
@@ -0,0 +1,30 @@
+
+#include "src/core/lib/security/credentials/s2a/s2a_credentials.h"
+
+static void test_s2a_credentials() { return; }
```
If there's a plan to implement this test function or add others, please add a TODO.

matthewstevenson88 replied:
Done.

cesarghali commented on test/core/security/s2a_security_connector_test.cc:
```diff
@@ -0,0 +1,30 @@
+
+#include "src/core/lib/security/security_connector/s2a/s2a_security_connector.h"
+
+static void test_s2a_security_connector() { return; }
```
Same here.

matthewstevenson88 replied:
Done.

matthewstevenson88 commented on include/grpc/grpc_security.h:
```diff
@@ -698,6 +698,52 @@ GRPCAPI grpc_channel_credentials* grpc_local_credentials_create(
+ *
+ *  On success, it returns the created S2A channel credential object; otherwise,
+ *  it returns nullptr. **/
+GRPCAPI grpc_channel_credentials* grpc_s2a_credentials_create(
```
This is often how the naming works in the gRPC C-core: for the client/channel credentials, its's simply `****_credentials`; for the server credentials, it's `****_server_credentials`. For example, search for the `alts_credentials_create` or `grpc_tls_spiffe_credentials_create` API's in [`grpc_security.h`](https://github.com/grpc/grpc/blob/master/include/grpc/grpc_security.h).

cesarghali replied:
SG, thanks!

matthewstevenson88 commented on src/core/lib/security/credentials/s2a/grpc_s2a_credentials_options.h:
```diff
@@ -0,0 +1,60 @@
+   * remove a duplicate ciphersuite or target service account from the
+   * appropriate vector, if a duplicate exists. **/
+  void set_handshaker_service_url(const char* handshaker_service_url);
+  void add_supported_ciphersuite(uint16_t ciphersuite);
```
They are defined in `s2a_constants.h`. I've added a reference to this.

matthewstevenson88 commented on src/core/lib/security/credentials/s2a/grpc_s2a_credentials_options.cc:
```diff
@@ -0,0 +1,116 @@
+    return;
+  }
+  char* service_account = gpr_strdup(target_service_account);
+  target_service_account_list_.push_back(service_account);
```
Maybe I misunderstood: are you asking why this method does not take ownership of `target_service_account`? If so, the answer is that I want to be able to free each element of `target_service_account_list_` in a consistent manner, and not have to worry about whether the caller created one dynamically or not.

cesarghali replied:
SG, thanks for the explanation.

matthewstevenson88 commented:
Thanks for your comments @cesarghali! I've either implemented or responded to each of them.

matthewstevenson88 commented:
Thanks for the review @cesarghali!

Squashing and merging now.
